Desktop e-book: Chapter 2 (#17300)
dotnet/docs: docs/architecture/modernize-desktop/whats-new-dotnet-core-desktop.md

---
title:  What's new with .NET Core 3.0 for Desktop?
description: .
ms.date: 09/16/2019
---

# What's new with .NET Core 3.0 for Desktop?

## From .NET Framework to .NET Core - The motivation behind .NET Core

Since its inception in 2002, .NET Framework has evolved through the years to support many technologies like Windows Forms, ASP.NET, Entity Framework, Windows Store, and many others. All of them are different in nature. Therefore, Microsoft was approaching this evolution by taking parts of the .NET Framework and creating a different application stack for each technology. That way, development capabilities could be customized for the needs of the specific stack, which maximized the potential of every platform. That lead to fragmentation on the versions of the .NET Framework maintained by different independent teams. All these stacks have a common structure, containing an App Model, a Framework, and a Runtime, but they differ in the implementation of each of these parts.

> ![Different Microsoft technologies](./media/whats-new-dotnet-core-desktop/different-technologies.png)

Author: Where did you see ASP.NET 5?

> ![Different Microsoft technologies](./media/whats-new-dotnet-core-desktop/different-technologies.png)

Reviewer: On the different-technologies.png image

- **Supported:** Microsoft officially supports .NET Core.

In .NET Core 3, besides existing web and cloud, there is also support for desktop, IoT, and AI domains. The goal for this framework is impressive: to target every type of .NET development present and future. Microsoft plans to complete this vision with .NET 5 at the end of 2020. The "Core" name was removed to reinforce its uniqueness in the .NET World.

![All domains of .NET 5](./media/whats-new-dotnet-core-desktop/all-domains-of-dotnet5.png)

## .NET Framework vs. .NET Core

Therefore, you now understand the relevance of .NET Core inside the Microsoft strategy for .NET and might be wondering what happens with .NET Framework and asking questions like: do you have to abandon it? Is it going to disappear? What are my choices to modernize the applications I have on .NET Framework?

The good news is Microsoft plans to support both .NET Framework and .NET Core. In fact, the announcement of .NET Core 3.0 was followed by exciting features in .NET Framework regarding modernization, so let's take a brief look at both paths.

> Listening to the .NET Framework developer's community suggestions, Microsoft have addressed three main scenarios for **.NET Core**:

Author: Rewrote

Regarding **.NET Framework 4.8**, Microsoft has addressed three scenarios:

- **Modern browser and media controls: Today, .NET desktop applications use Internet Explorer and Windows Media Player for showing HTML and playing media files. Since these legacy controls don't show the latest HTML or play the latest media files, we are adding new controls that take advantage of Microsoft Edge and newer media players to support the latest standards.
- **Access to UWP Controls: UWP contains new controls that take advantage of the latest Windows features and touch displays. You will not have to rewrite your applications to use these new features and controls. Now they are available to Windows Forms and WPF so that you can take advantage of these new features in your existing code.
- **High DPI improvements: The resolution of displays is increasing to 4K and 8K resolutions. We want to make sure your existing Windows Forms and WPF applications can look great on these new displays. That's why we are adding improvements for HDPI

> Since .NET Framework is working on millions of machines, Microsoft will continue to support it but will not be adding new features.

Reviewer: I don't think this statement is the real reason why we support it.

```suggestion
Since .NET Framework is installed on millions of machines, Microsoft will continue to support it but won't add new features.
```

> .NET Core is the open source, cross-platform, and fast-moving version of .NET. Because of its side-by-side nature, it can take changes without the fear of breaking any application. This means that .NET Core will get new APIs and language features over time that .NET Framework cannot.

Reviewer: ```suggestion
.NET Core is the open-source, cross-platform, and fast-moving version of .NET. Because of its side-by-side nature, it can take changes without the fear of breaking any application. This means that .NET Core will get new APIs and language features over time that .NET Framework won't.
```

> 

Reviewer: Again, I don't think this statement above is accurate. I think the reason why Framework won't be getting new APIs and language features is because we decided to invest in one platform only. So I don't see the casualty between the sxs nature of .NET Core and Framework not being able to get those.

> An essential part of the roadmap is to ease developers to move applications to .NET Core. .NET Core 3.0 takes a huge step by adding WPF, Windows Forms and Entity Framework 6 support, and we will keep porting APIs and features to help close the gap and make migration easier for those who chose to do so.

Reviewer: ```suggestion
An essential part of the roadmap is to ease developers to move applications to .NET Core. .NET Core 3.0 takes a huge step by adding WPF, Windows Forms, and Entity Framework 6 support, and we'll keep porting APIs and features to help close the gap and make migration easier for those developers who chose to do so.
```

> So, if you have existing .NET Framework applications, you should not feel pressured to move to .NET Core. .NET Framework will be fully supported and will always be a part of Windows. However, if in the future you want to use the newest language features and APIs, you'll need to move your applications to .NET Core.

Reviewer: is that statement true? That .NET Framework will always be a part of Windows?

```suggestion
If you have existing .NET Framework applications, you shouldn't feel pressured to move to .NET Core. .NET Framework will be fully supported and will always be a part of Windows. However, if you want to use the newest language features and APIs in the future, you'll need to move your applications to .NET Core.
```

> For your brand-new desktop applications, we recommend starting directly on .NET Core. It is lightweight, runs side-by-side, high performance and cross-platform and fits perfectly on containers and microservices architectures.

Reviewer: ```suggestion
For your brand new desktop applications, we recommend starting directly on .NET Core. It's lightweight and cross platform, runs side by side, has high performance, and fits perfectly on containers and microservices architectures.
```

> ![.NET Framework and .NET Core](./media/whats-new-dotnet-core-desktop/dotnet-framework-and-dotnet-core.png)

Reviewer: improve alt text and maybe enlarge the image? It's hard to read.
also avoid dupe words on the path. maybe framework-vs-core.png

## .NET Standard vs. PCL

> The [.NET Standard](https://github.com/dotnet/standard) is a formal specification of .NET APIs that are intended to be available on all .NET implementations. The motivation behind the .NET Standard is establishing greater uniformity in the .NET ecosystem. .NET Standard is a specification of .NET APIs that make up a uniform set of contracts to compile your code against. These contracts are implemented in each .NET flavor, thus enabling portability across different .NET implementations.

Reviewer: why link to the repo here and to the article before?

The .NET Standard enables the following key scenarios:

- Defines uniform set of base class libraries APIs for all .NET implementations to implement, independent of the workload.
- Enables developers to produce portable libraries that are usable across .NET implementations, using this same set of APIs.

> - Reduces or even eliminates conditional compilation of shared source due to .NET APIs, only for OS APIs.

Reviewer: not sure if I follow this one.

Author: Removed it

.NET Standard is the evolution of PCLs and here are the fundamental differences between .NET Standard and PCLs:

> - .NET Standard is a set of curated APIs, picked by Microsoft, PCLs are not.

Reviewer: ```suggestion
- .NET Standard is a set of curated APIs, picked by Microsoft. PCLs aren't.
```

- The APIs that a PCL contains are dependent on the platforms that you choose to target when you create a PCL. This makes a PCL only sharable for the specific targets that you choose.
- .NET Standard is platform-agnostic, it can run anywhere, on Windows, macOS, Linux, and so on.
- PCLs can also run cross-platform, but they have a more limited reach. PCLs can only target a limited set of platforms.

## .NET Core 3.0 new Desktop features

### Support for Windows Forms and WPF

> Windows Forms and WPF are part of .NET Core 3.0. Both presentation frameworks are only for the Windows OS and they are not cross-platform. You can think of WPF as a rich layer over DirectX and Windows Forms as a thinner layer over GDI+. WPF and Windows Forms do a great job of exposing and exercising much of the desktop application functionality in Windows. With .NET Core 3.0. Windows Forms and WPF become available for .NET Core as well as for .NET Framework. Now you can start your new desktop applications targeting .NET Core and migrate your existing ones from .NET Framework to .NET Core.

Reviewer: ```suggestion
Windows Forms and WPF are part of .NET Core 3.0. Both presentation frameworks are for Windows only, so they aren't cross platform. You can think of WPF as a rich layer over DirectX and Windows Forms as a thinner layer over GDI+. WPF and Windows Forms do a great job of exposing and exercising much of the desktop application functionality in Windows. With .NET Core 3.0. Windows Forms and WPF are available for .NET Core and .NET Framework. Now you can start your new desktop applications targeting .NET Core and migrate your existing ones from .NET Framework to .NET Core.
```

A new version of .NET Standard, version 2.1, was released at the same time. As expected, all new .NET Standard APIs are part of .NET Core 3.0.

> You can build desktop applications with C#, F#, and VB with .NET Core 3.0.

Reviewer: did you ship the templates for all of these? If not, it would be good to call out, no?

Author: Removed the line because right now only C# makes sense, and we talked about desktop apps available on Core already

> Also, important to notice that both Windows Forms and WPF implementations for .NET Core 3 are being open-sourced.

Reviewer: ```suggestion
Also, it's important to notice that both Windows Forms and WPF implementations for .NET Core 3.0 are being open sourced.
```

### XAML Islands

> 

Reviewer: should you link to some articles that talk more about XAML islands?

XAML Islands is a set of components for developers to use the new Windows 10 controls (UWP XAML controls) in their current WPF, Windows Forms, and native Win32 apps (like MFC). You can have your "islands" of UWP XAML controls wherever you want inside your Win32 apps.

> These XAML Islands are possible because Windows 10 1903 update introduces a set of APIs that allows hosting UWP XAML content in Win32 windows using windows handlers (HWnds). Notice that only apps running on Windows 10 1903 and above can use XAML Islands.

Reviewer: ```suggestion
These XAML Islands are possible because Windows 10, version 1903 introduces a set of APIs that allows hosting UWP XAML content in Win32 windows using windows handlers (HWnds). Notice that only apps running on Windows 10 1903 and above can use XAML Islands.
```

> To facilitate to create XAML Islands for Windows Forms and WPF developers, the Windows Community Toolkit introduces a set of .NET wrappers in several NuGet packages. Those wrappers are the Wrapped Controls and Hosting Controls:

Reviewer: ```suggestion
To make it easier to create XAML Islands for Windows Forms and WPF developers, the Windows Community Toolkit introduces a set of .NET wrappers in several NuGet packages. Those wrappers are the wrapped and hosting controls:
```

> - The WebView, WebViewCompatible, [InkCanvas](/windows/communitytoolkit/controls/wpf-winforms/inkcanvas), [MediaPlayerElement](/windows/communitytoolkit/controls/wpf-winforms/mediaplayerelement), and [MapControl](/windows/communitytoolkit/controls/wpf-winforms/mapcontrol) wrapped controls wrap some UWP XAML controls into Windows Forms or WPF controls, hiding UWP concepts for those developers.

Reviewer: ```suggestion
- The [WebView](/windows/communitytoolkit/controls/wpf-winforms/webview), [WebViewCompatible](/windows/communitytoolkit/controls/wpf-winforms/webviewcompatible), [InkCanvas](/windows/communitytoolkit/controls/wpf-winforms/inkcanvas), [MediaPlayerElement](/windows/communitytoolkit/controls/wpf-winforms/mediaplayerelement), and [MapControl](/windows/communitytoolkit/controls/wpf-winforms/mapcontrol) wrapped controls wrap some UWP XAML controls into Windows Forms or WPF controls, hiding UWP concepts for those developers.
```

> - The WindowsXamlHost control for Windows Forms and WPF allows others not-wrapped UWP XAML controls and custom controls can be loaded into a XAML Island.

Reviewer: ```suggestion
- The [WindowsXamlHost](/windows/communitytoolkit/controls/wpf-winforms/windowsxamlhost) control for Windows Forms and WPF allows others not-wrapped UWP XAML controls and custom controls can be loaded into a XAML Island.
```

When a new .NET Core version is released, you can update each app on a machine as needed without any concern of affecting other applications. New .NET Core versions are installed in their own directories and exist "side-by-side" with each other.

If you need to deploy with isolation, you can deploy .NET Core with your application. .NET Core will bundle your app with the .NET Core runtime as in a single executable.

These deployment options were requested by developers for quite a long time but were difficult to achieve using .NET Framework. The modular architecture used by .NET Core makes these flexible deployment options possible.

### Performance

Since its inception, targeting the Web and Cloud workloads, .NET Core has had performance plugged into its DNA. Server-side code must be performant enough to fulfill high concurrency scenarios and .NET Core scores today as the best performance web platform in the market.

> You can now take advantage of these performance improvements when you use it to build your next generation of desktop applications.

Reviewer: ```suggestion
You can take advantage of these performance improvements when you use .NET Core to build your next generation of desktop applications.
```

## Benefits of open-source

Just a few words about .NET Core being open source. Building a cross-platform stack is something complex that needs the interaction of specialized teams on each of the targeted platforms. This effort needs much collaboration from inside and outside of Microsoft. By making it open source and thus open to public collaboration, you get the ultimate agile development style in place, raising the quality bar since issues are detected by a huge and active community of developers.

> This is a key success factor of .NET Core that will continue to speed up the roadmap mentioned above: To be the single .NET platform that any developer will ever need to build any application.

Reviewer: ```suggestion
This is a key success factor of .NET Core that will continue to speed up the roadmap previously mentioned: To be the single .NET platform that any developer will ever need to build any application.
```
